File: src/main/java/com/homeclimatecontrol/esphome2influxdb/Worker.java
package com.homeclimatecontrol.esphome2influxdb;

import java.util.concurrent.CountDownLatch;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public abstract class Worker<T extends Endpoint> implements Runnable {

    protected final Logger logger = LogManager.getLogger();

    /**
     * Endpoint to connect to.
     */
    protected final T endpoint;

    /**
     * Gate to {@code countDown()} when the worker shutdown is complete.
     */
    protected final CountDownLatch stoppedGate;

    protected Worker(T endpoint, CountDownLatch stoppedGate) {
        this.endpoint = endpoint;
        this.stoppedGate = stoppedGate;
    }
}

File: src/main/kotlin/com/homeclimatecontrol/esphome2influxdb/k/Worker.kt
package com.homeclimatecontrol.esphome2influxdb.k

import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.util.concurrent.CountDownLatch

/**
 * @param endpoint Endpoint to connect to.
 * @param stoppedGate Gate to `countDown()` when the worker shutdown is complete.
 */
abstract class Worker<T : Endpoint>(
    protected val endpoint: T,
    protected val stoppedGate: CountDownLatch
) {
    protected val logger: Logger = LogManager.getLogger()
    public abstract suspend fun run();
}
